caas/kubernetes/provider/k8s.go

```diff
@@ -1813,14 +1856,17 @@ func (k *kubernetesClient) configureHeadlessService(
 	logger.Debugf("creating/updating headless service for %s", appName)
 	service := &core.Service{
 		ObjectMeta: v1.ObjectMeta{
-			Name:   headlessServiceName(deploymentName),
-			Labels: utils.LabelsForApp(appName),
+			Name: headlessServiceName(deploymentName),
+			Labels: utils.LabelsMerge(
+				utils.LabelsForApp(appName, k.IsLegacyLabels()),
+				utils.LabelsJuju,
+			),
 			Annotations: k8sannotations.New(nil).
 				Merge(annotations).
 				Add("service.alpha.kubernetes.io/tolerate-unready-endpoints", "true").ToMap(),
 		},
 		Spec: core.ServiceSpec{
-			Selector:                 utils.LabelsForApp(appName),
+			Selector:                 utils.LabelsForApp(appName, k.IsLegacyLabels()),
```

Comment:
Oh I see, sometimes we do want just these.
Perhaps then have SelectorLabelsForApp() and a LabelsForApp() methods to avoid the caller needed to figure out what to do

Reply:
It's not that easy. We could build out generic routines for all the object types that exist